sdk/include/opentelemetry/sdk/metrics/state/attributes_hashmap.h

```diff
@@ -0,0 +1,130 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifndef ENABLE_METRICS_PREVIEW
+#  include "opentelemetry/common/spin_lock_mutex.h"
+#  include "opentelemetry/nostd/function_ref.h"
+#  include "opentelemetry/version.h"
+
+#  include <functional>
+#  include <memory>
+#  include <mutex>
+#  include <unordered_map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+using opentelemetry::sdk::common::OrderedAttributeMap;
+
+class AttributeHashGenerator
+{
+public:
+  size_t operator()(const MetricAttributes &attributes) const
+  {
+    return opentelemetry::sdk::common::GetHashForAttributeMap(attributes);
+  }
+};
+
+class AttributesHashMap
+{
+public:
+  Aggregation *Get(const MetricAttributes &attributes) const
+  {
+    std::lock_guard<opentelemetry::common::SpinLockMutex> guard(GetLock());
+    auto it = hash_map_.find(attributes);
+    if (it != hash_map_.end())
+    {
+      return it->second.get();
+    }
+    return nullptr;
+  }
+
+  /**
+   * @return check if key is present in hash
+   *
+   */
+  bool Has(const MetricAttributes &attributes) const
+  {
+    std::lock_guard<opentelemetry::common::SpinLockMutex> guard(GetLock());
+    return (hash_map_.find(attributes) == hash_map_.end()) ? false : true;
+  }
+
+  /**
+   * @return the pointer to value for given key if present.
+   * If not present, it uses the provided callback to generate
+   * value and store in the hash
+   */
+
+  Aggregation *GetOrSetDefault(const MetricAttributes &attributes,
+                               std::function<std::unique_ptr<Aggregation>()> aggregation_callback)
+  {
+    std::lock_guard<opentelemetry::common::SpinLockMutex> guard(GetLock());
+    auto it = hash_map_.find(attributes);
+    if (it != hash_map_.end())
+    {
+      return it->second.get();
+    }
+
+    hash_map_[attributes] = std::move(aggregation_callback());
+    return hash_map_[attributes].get();
+  }
+
+  /**
+   * Set the value for given key, overwriting the value if already present
+   */
+
+  void Set(const MetricAttributes &attributes, std::unique_ptr<Aggregation> value)
```

Comment:
The comment doesn't match with the code.

Reply:
Good catch, comments are correct have fixed the code to match that :)